src/librustc_typeck/check/pat.rs

```diff
@@ -1221,15 +1252,62 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                     it explicitly.",
             );
         }
-        err.emit();
+        err
+    }
+
+    fn error_tuple_variant_as_struct_pat(
+        &self,
+        pat: &Pat<'_>,
+        fields: &'tcx [hir::FieldPat<'tcx>],
+        variant: &ty::VariantDef,
+    ) -> Option<DiagnosticBuilder<'tcx>> {
+        if let (CtorKind::Fn, PatKind::Struct(qpath, ..)) = (variant.ctor_kind, &pat.kind) {
+            let path = rustc_hir_pretty::to_string(rustc_hir_pretty::NO_ANN, |s| {
```

Comment:
It's better to avoid the expensive error creation in successful case.
Matching `Tuple(Something)` with `Tuple { .. }` may could be common enough, especially if auto-generated.
The error can be created only if we actually hit the `(Some(mut err), None, None)` case.